evm_loader/program/src/evm/database.rs

```diff
@@ -51,3 +49,244 @@ pub trait Database {
         is_static: bool,
     ) -> Option<Result<Vec<u8>>>;
 }
+
+/// Provides convenience methods that can be implemented in terms of `Database`.
+#[maybe_async(?Send)]
+pub trait DatabaseExt {
+    /// Returns whether an account exists, which is also referred to as the account being non-empty,
+    /// in accordance with https://eips.ethereum.org/EIPS/eip-161.
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool>;
+
+    /// Returns the code has for an address in accordance with https://eips.ethereum.org/EIPS/eip-1052.
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]>;
+}
+
+#[maybe_async(?Send)]
+impl<T: Database> DatabaseExt for T {
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool> {
+        Ok(self.nonce(address, chain_id).await? > 0 || self.balance(address, chain_id).await? > 0)
+    }
+
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]> {
+        // The function `Database::code` returns a zero-length buffer if the account exists with
+        // zero-length code, but also when the account does not exist. This makes it necessary to
+        // also check if the account exists when the returned buffer is empty.
+        //
+        // We could simplify the implementation by checking if the account exists first, but that
+        // would lead to more computation in what we think is the common case where the account
+        // exists and contains code.
+        let code = self.code(address).await?;
+        let bytes_to_hash: Option<&[u8]> = if !code.is_empty() {
+            Some(&*code)
+        } else if self.account_exists(address, chain_id).await? {
+            Some(&[])
+        } else {
+            None
+        };
+
+        Ok(bytes_to_hash.map_or([0; 32], |bytes| {
+            solana_program::keccak::hash(bytes).to_bytes()
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
This is a duplicate of already existing tests - https://github.com/neonlabsorg/neon-tests/blob/develop/integration/tests/basic/evm/opcodes/test_extcodehash.py

Also you are not validating anything useful here. Instead you should add additional integration tests that cover the use case you are trying to fix. Ask QA for the help.

Reply:
@kristinaNikolaevaa fyi

Reply:
These unit tests validate the implementation logic of `DatabaseExt::code_hash`. 

We should have the integration tests indeed. Working on that with @kristinaNikolaevaa.

Are you suggesting we don't use unit tests? If so, I will remove them. Just wanted to confirm.

Reply:
@anton-lisanin I think we should leave the unit tests too, because they are also a hint about making the code more testable.

The added tests might look weird, but that big `Database` trait with many non-implemented methods could be broken down into smaller ones in another PR.

The `tokio` as dev-dependency part is needed because when building `evm_loader` crate as a dependency of `neon-lib` crate (as part of `cargo test`), it is built in async mode.

Reply:
Conclusion from daily was to merge as is and discuss unit testing soon.